packages/aws-cdk/lib/cdk-toolkit.ts

```diff
@@ -534,6 +541,91 @@ export class CdkToolkit {
     return stacks;
   }
 
+  private stackHasNestedStacks(rootStack: cxapi.CloudFormationStackArtifact) {
+    for (const resourceName in rootStack.template.Resources) {
+      if (rootStack.template.Resources[resourceName].Type === 'AWS::CloudFormation::Stack') {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private async findPhysicalNameForNestedStack(logicalId: string, parentStackName: string | undefined, sdk: ISDK): Promise<string | undefined> {
+    if (!parentStackName) {
+      return undefined;
+    }
+    const stackResources = await this.listNestedStackResources(parentStackName, sdk);
+    return stackResources.find(sr => sr.LogicalResourceId === logicalId)?.PhysicalResourceId;
+  }
+
+  private async listNestedStackResources(parentStackName: string, sdk: ISDK): Promise<CloudFormation.StackResourceSummary[]> {
+    const ret = new Array<CloudFormation.StackResourceSummary>();
+    let nextToken: string | undefined;
+    do {
+      const stackResourcesResponse = await sdk.cloudFormation().listStackResources({
+        StackName: parentStackName,
+        NextToken: nextToken,
+      }).promise();
+      ret.push(...(stackResourcesResponse.StackResourceSummaries ?? []));
+      nextToken = stackResourcesResponse.NextToken;
+    } while (nextToken);
+    return ret;
+  }
```

Comment:
This is all taken from the hotswapping logic ([here](https://github.com/aws/aws-cdk/blob/9e6f977d8ac4ad7ab2852536cc20c4469fe74f03/packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts#L48-L51) and [here](https://github.com/aws/aws-cdk/blob/9e6f977d8ac4ad7ab2852536cc20c4469fe74f03/packages/aws-cdk/lib/api/hotswap-deployments.ts#L247-L266)).

We need to extract it out, and get rid of this duplication.